topology/nhlt/intel/dmic/dmic-process.c

```diff
@@ -0,0 +1,1341 @@
+				dmic->dmic_fir_array.fir_coeffs[i][di][j] = cu;
+			}
+			dmic->dmic_fir_array.fir_len[0] = length;
+			dmic->dmic_fir_array.fir_len[1] = 0;
+		}
+
+		if (di == 1) {
+			/* FIR B */
+			fir_decim = MAX(cfg->mfir_b - 1, 0);
+			fir_length = MAX(cfg->fir_b_length - 1, 0);
+			val = FIR_CONTROL_B_START(1) |
+				FIR_CONTROL_B_ARRAY_START_EN(array_b) |
+				FIR_CONTROL_B_DCCOMP(dccomp) |
+				FIR_CONTROL_B_MUTE(fir_mute) |
+				FIR_CONTROL_B_STEREO(stereo[i]);
+			dmic->dmic_blob_fir[i][di].fir_control = val;
+
+			val = FIR_CONFIG_B_FIR_DECIMATION(fir_decim) |
+				FIR_CONFIG_B_FIR_SHIFT(cfg->fir_b_shift) |
+				FIR_CONFIG_B_FIR_LENGTH(fir_length);
+			dmic->dmic_blob_fir[i][di].fir_config = val;
+			val = DC_OFFSET_LEFT_B_DC_OFFS(DCCOMP_TC0);
+			dmic->dmic_blob_fir[i][di].dc_offset_left = val;
+
+			val = DC_OFFSET_RIGHT_B_DC_OFFS(DCCOMP_TC0);
+			dmic->dmic_blob_fir[i][di].dc_offset_right = val;
+
+			val = OUT_GAIN_LEFT_B_GAIN(0);
+			dmic->dmic_blob_fir[i][di].out_gain_left = val;
+
+			val = OUT_GAIN_RIGHT_B_GAIN(0);
+			dmic->dmic_blob_fir[i][di].out_gain_right = val;
+
+			/* Write coef RAM B with scaled coefficient in reverse order */
+			length = cfg->fir_b_length;
+			for (j = 0; j < length; j++) {
+				ci = (int32_t)Q_MULTSR_32X32((int64_t)cfg->fir_b->coef[j],
+							     cfg->fir_b_scale, 31,
+							     DMIC_FIR_SCALE_Q, DMIC_HW_FIR_COEF_Q);
+				cu = FIR_COEF_B(ci);
+				/* blob_pdm[i].fir_coeffs[di][j] = cu; */
+				dmic->dmic_fir_array.fir_coeffs[i][di][j] = cu;
+			}
+			dmic->dmic_fir_array.fir_len[1] = length;
+		}
+	}
+
+	return 0;
+}
+
+/* The decimation for PDM (pulse density modulation) stream is done in a programmable HW filter
+ * engine. The input to configuration algorithm is needed sample rate, channels/enabled microphones,
+ * microphone clock range, microphone clock duty cycle range, and system clock rate.
+ *
+ * The PDM bus clock divider, CIC and FIR decimation ratios are searched and configuration for
+ * optimal power consumption, filtering requirements, and HW constraints is chosen. The FIR filter
+ * for the chosen decimation is looked up from table and scaled to match the other decimation path
+ * sensitivity.
+ */
+int dmic_calculate(struct intel_nhlt_params *nhlt)
+{
+	struct intel_dmic_params *dmic = (struct intel_dmic_params *)nhlt->dmic_params;
+	struct dmic_calc_matched_modes modes_ab;
+	struct dmic_calc_decim_modes modes_a;
+	struct dmic_calc_decim_modes modes_b;
+	struct dmic_calc_configuration cfg;
+	int ret = 0;
+	int di;
+
+	if (!dmic)
+		return -EINVAL;
+
+	di = dmic->dmic_dai_index;
+
+	if (di >= DMIC_HW_FIFOS) {
+		fprintf(stderr, "dmic_set_config(): dai->index exceeds number of FIFOs\n");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (dmic->dmic_prm[di].num_pdm_active > DMIC_HW_CONTROLLERS) {
+		fprintf(stderr, "dmic_set_config():controller count exceeds platform capability\n");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	switch (dmic->dmic_prm[di].fifo_bits) {
+	case 0:
```

Comment:
is zero not an error?

Reply:
according to @singalsu this is used to indicate that the fifo is not in use... so I leave it that way